include/unifex/indexed_for.hpp

```diff
@@ -151,29 +318,58 @@ struct indexed_for_sender {
     friend void tag_invoke(
         tag_t<visit_continuations>,
-        const indexed_for_receiver& r,
+        const double_indexed_for_receiver& r,
         Visit&& visit) {
       std::invoke(visit, r.receiver_);
     }
+
   };
 
   template <typename Receiver>
   auto connect(Receiver&& receiver) && {
+    // Connect the receiver to the predecssor of pred_
+    // we skip the connect/start pairing on the immediate
+    // predecessor as an optimisation (or at least, to
+    // emulate the optimisation we might see in
+    // accelerator implementations)
     return unifex::connect(
-        std::forward<Predecessor>(pred_),
-        indexed_for_receiver<std::remove_cvref_t<Receiver>>{
+        std::forward<Predecessor>(pred_).pred_,
+        double_indexed_for_receiver<std::remove_cvref_t<Receiver>>{
+            std::forward<Predecessor>(pred_).func_,
             std::forward<Func>(func_),
+            std::forward<Predecessor>(pred_).policy_,
             std::forward<Policy>(policy_),
+            std::forward<Predecessor>(pred_).range_or_selector_,
             std::forward<RangeOrRangeSelector>(range_or_selector_),
             std::forward<Receiver>(receiver)});
   }
 };
 
-template <typename Sender, typename Policy, typename RangeOrRangeSelector, typename Func>
-auto indexed_for(Sender&& predecessor, Policy&& policy, RangeOrRangeSelector&& range, Func&& func) {
-  return indexed_for_sender<
-      std::remove_cvref_t<Sender>, std::decay_t<Policy>, std::decay_t<RangeOrRangeSelector>, std::decay_t<Func>>{
-      (Sender &&) predecessor, (Policy&&) policy, (RangeOrRangeSelector&& ) range, (Func &&) func};
+// Customisation for when two indexed_for_senders are chained
+// This simulates how we might customise an arbitrary set of types for
+// optimal interaction
+template <typename InnerPredecessor, typename InnerPolicy, typename InnerRange, typename InnerFunc, typename Policy, typename Range, typename Func>
+auto
+tag_invoke(
+    indexed_for_customization_cpo,
+    indexed_for_sender<InnerPredecessor, InnerPolicy, InnerRange, InnerFunc>&& predecessor,
+    Policy&& policy, Range&& range, Func&& func) noexcept {
+
+  // Construct the doubled version of the sender that grabs content from the predecessor directly
+  // and does not apply connect/start in between.
+  return double_indexed_for_sender<
+        std::remove_cvref_t<indexed_for_sender<InnerPredecessor, InnerPolicy, InnerRange, InnerFunc>>,
+        std::decay_t<Policy>, std::decay_t<Range>, std::decay_t<Func>>{
+      (indexed_for_sender<InnerPredecessor, InnerPolicy, InnerRange, InnerFunc>&& ) predecessor,
+      (Policy&&) policy,
+      (Range&& ) range,
+      (Func &&) func};
+}
+
+template <typename Sender, typename Policy, typename Range, typename Func>
+auto indexed_for(Sender&& predecessor, Policy&& policy, Range&& range, Func&& func) {
+  return indexed_for_customization(
+      (Sender &&) predecessor, (Policy&&) policy, (Range&& ) range, (Func &&) func);
 }
```

Comment:
I don't think we need this function. We can just rename the `indexed_for_customization` CPO to `indexed_for`.